chore(*): realign master branch to 3.0.0 release

nowNick opened the pull request:
## Description 

This PR fixes master branch so that it points to the latest changes and to the release of 3.0.0 introduced here: https://github.com/Kong/lua-resty-healthcheck/pull/142

It reverts the master branch back to 1.3.0 release (which is: dc2a6b60384f22be984be6e29090e9ad8c631fee) via commit: d4eac10e79c2512866b377a4a5a9cf5fb406beb5 and then reapplies the changes introduced in 1.4.x, 1.5.x, 1.6.x and 3.0.0.

## Issue reference:
[KAG-2704](https://konghq.atlassian.net/browse/KAG-2704)

[KAG-2704]: https://konghq.atlassian.net/browse/KAG-2704?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

HansK-p committed:
Added support for https_sni in healthcheck.lua (#49)

Tieske committed:
feat(*) add header support for active checks

locao committed:
release 1.4.0

locao committed:
fix(healthcheck) use single timer for all active checks (#62)

* fix(healthcheck) use single timer for all active checks

* tests(*) removed tests that are not needed

locao committed:
docs(*) docs for release 1.4.0

locao committed:
chore(ci) use newer openresty and luarocks releases (#68)

locao committed:
fix(healthcheck) single worker actively checks the status (#67)

locao committed:
release 1.4.1

onematchfox committed:
fix(healthcheck) record `last_run` when healthcheck is scheduled (#72)

Prevents a thundering herd issue whereby additional healthchecks are scheduled in the time in which it takes the healthcheck to complete.

Commit:
tests(active-probes) interval is respected (#73)

* fix(healthcheck) record `last_run` when healthcheck is scheduled

Prevents a thundering herd issue whereby additional healthchecks are scheduled in the time in which it takes the healthcheck to complete.

* tests(active-probes) interval is respected

Co-authored-by: Brian Fox <brianhfox@gmail.com>

locao committed:
fix(healthcheck) remove event watcher when stopping hc (#74)

Co-authored-by: Brian Fox <brianhfox@gmail.com>

Co-authored-by: Brian Fox <brianhfox@gmail.com>

locao committed:
tests(*) avoid some flakiness (#75)

locao committed:
release 1.4.2

locao committed:
chore(*) add GitHub Actions workflows (#82)

* chore(*) add GitHub Actions workflows

* fix(healthcheck) lint error

javierguerragiraldez committed:
Simplify start of the checking timer (#85)

* simplify start of the checking timer, ensuring only one worker actively sends healthchecks.

one timer per worker, but before doing anything, tries to acquire an
expiration lock.  if fails, try again later.  if the "winning" worker
ever fails to renew it, some other worker would get it.

locao committed:
chore(rockspec) added rockspec for release 1.5.0-1

Also:
  - updated scm-1 rockspec
  - bumped openresty version in CI tests

mayocream committed:
feat(active) support map headers

locao committed:
feat(healthcheck) delayed_clear function (#88)

Added new function delayed_clear. This function marks all targets to be
removed, but do not actually remove them. If before the delay parameter
any of them is re-added, it is unmarked for removal.

This function makes it possible to keep target state during config
changes, where the targets might be removed and then re-added.

locao committed:
chore(readme) 1.5.0 release (#91)

* chore(readme) 1.5.0 release

* docs(*) release 1.5.0

Also added docs missing to delayed_clear() function.

kayano committed:
fix(healthcheck) Use pair instead ipair for hcs weak table (#93)

locao committed:
release 1.5.1 (#95)

locao committed:
chore(readme) update badges (#98)

locao committed:
docs(readme) updated with 1.4.x changes

ADD-SP committed:
fix(mtls) use OpenResty's API for mtls (#99)

locao committed:
chore(workflows) updates for 1.6.0 release

- added latest openresty to the CI matrix
- added tests for when lua-resty-worker-events or lua-resty-events are
  used

locao committed:
feat(healthcheck) support setting the events module (#105)

* feat(healthcheck) support setting the events module

* fix(healthcheck) defaults to lua-resty-worker-events

* tests(workflows) fixed manual deps install

* fix(healthcheck) check empty opts

* chore(workflows) use last luarocks

* test(workflows) use pre-built deps, test with or 1.13-1.21

* chore(workflows) install lua-resty-events in ci

* tests(workflows) debug

* fixed tests and resty-events usage

* init resty-events in init_worker

* fix(tests) init events module (#107)

* add init_worker in 03-get_target_status.t

* fix 03-get_target_status.t

* fix 03-get_target_status_with_sleeps.t

* fix 04-report_success.t

* fix 05/06

* fix 07/08

* fix 09

* change 10

* fix 11

* fix 12

* change 13

* fix 15

* partial fix 16

* change 17

* fix 18

* change 13

* fix 16

* style 05

* fix 01/02

* use string.buffer in OpenResty 1.21.4.1 (#109)

* use string.buffer in OpenResty 1.21.4.1

* remove cjson require

* fix(healthcheck) use the events module set in defaults

* tests(with_resty-events) disabled tests that need more work

* fix(healthcheck) avoid breaking when opts are nil

* tests(with_resty-events) removed unnecessary test

* tests(with_resty-events) increased sleeps

Co-authored-by: Chrono <chrono_cpp@me.com>

locao committed:
release 1.6.0 (#110)

* docs(readme) release 1.6.0

* fix(rockspec) typo

* chore(rockspec) release 1.6.0

* docs(*) release 1.6.0

chronolaw committed:
chore(*) localize string.format (#111)

flrgh committed:
fix(healthchecker) port 2.x lock fixes to 1.5.x (#113)

* fix(healthchecker) port 2.x lock fixes to 1.5.x

* chore(healthcheck) remove unused vars

* chore(healthcheck) fix indent level

* fix(healthcheck) correct duplicate handling in add_target

* fix(healthchecker) handle fetch_target_list failure in checker callback

* chore(healthcheck) apply suggestions from #112

Co-authored-by: Vinicius Mignot <vinicius.mignot@gmail.com>

locao committed:
fix(healthcheck) lower the cleanup check frequency

the health-check timer also checks if targets must be removed. to
safely remove targets, the targets list is locked. if this check runs
on every health-check cycle and there are a large number of targets, a
bazillion locks will be created. this change avoids that by lowering
the frequency the cleanup list is checked.

the side-effect is that targets marked for cleanup may exist for more
time (2.5s) than expected, and some unexpected active checks could
happen.

flrgh committed:
chore(healthcheck) increase verbosity for locked function failures (#114)

* chore(healthcheck) increase verbosity for locked function failures

* tests(healthcheck) add tests for run_locked()

locao committed:
tests(clear) increase delay for delayed clear tests

with less locks the wait for delayed clean is longer.

locao committed:
fix(healthcheck) support any lua-resty-events 0.1.x (#118)

* chore(workflows) bump deps versions

* chore(helathcheck) support any lua-resty-events 0.1.x

locao committed:
docs(readme) release 1.6.1

locao committed:
chore(rockspecs) release 1.6.1

locao committed:
release 1.6.1

locao committed:
docs(readme) updated build badge

oowl committed:
feat(healthcheck) avoid duplication post in rebuild healthcheck scenario

locao committed:
chore(ci) remove old openresty versions

locao committed:
release 1.6.2

fffonion committed:
chore(ci): fix cache path (#136)

${{ env.* }} is not evaluated in `with` causing gha tries to cache `/`.

bungle committed:
release 1.6.3 (#135)

nowNick committed:
release 3.0.0 (#142)

* feat(ci/KAG-1800): add lint and sast workflows using shared actions

* chore(ci): pin shared code quality actions

* chore(*): backport - localize some functions

A commit on master 80ee2e170dc89ab824226ca6d806c4ab6f7f4cdb
introduced localizing some functions. This commit backports that one.

Backports: https://github.com/Kong/lua-resty-healthcheck/pull/92

* fix(healthcheck): fixed incorrect default http_statuses when new() was called multiple times (#83)

* chore(lint): bump kong/public-shared-actions

* docs(README): added 1.5.2 and 1.5.3 releases

* chore(*) rename readme, add release instructions

* chore(healthcheck): fix get_defaults function

* fix(test): fix worker-events test

* release 3.0.0

* chore(github): cancel in progress workflows when new pushed

---------

Co-authored-by: saisatish karra <saisatish.karra@konghq.com>
Co-authored-by: Shuoqing Ding <dsq704136@gmail.com>
Co-authored-by: Vinicius Mignot <vinicius.mignot@gmail.com>
Co-authored-by: Thijs Schreijer <thijs@thijsschreijer.nl>

nowNick committed:
chore(*): revert commits back to 1.3.0

This reverts the master branch backs to the commit
of dc2a6b60384f22be984be6e29090e9ad8c631fee so that
we can skip over 2.0.0 release.
The 1.3.0 release is the first common commit between
master branch and 1.6.x (also 3.0.x) branches.

nowNick committed:
Merge branch 'release/3.0.x' into chore/revert-master-branch

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/lua-resty-healthcheck?pullRequest=144) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Kong/lua-resty-healthcheck?pullRequest=144) before we can accept your contribution.<br/>**8** out of **10** committers have signed the CLA.<br/><br/>:white_check_mark: onematchfox<br/>:white_check_mark: kayano<br/>:white_check_mark: chronolaw<br/>:white_check_mark: oowl<br/>:white_check_mark: flrgh<br/>:white_check_mark: ADD-SP<br/>:white_check_mark: mayocream<br/>:white_check_mark: nowNick<br/>:x: Murillo Paula<br/>:x: HansK-p<br/><hr/>**Murillo Paula** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/lua-resty-healthcheck?pullRequest=144) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Kong/lua-resty-healthcheck?pullRequest=144) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Kong/lua-resty-healthcheck?pullRequest=144) before we can accept your contribution.<br/>**8** out of **10** committers have signed the CLA.<br/><br/>:white_check_mark: onematchfox<br/>:white_check_mark: kayano<br/>:white_check_mark: mayocream<br/>:white_check_mark: chronolaw<br/>:white_check_mark: oowl<br/>:white_check_mark: ADD-SP<br/>:white_check_mark: nowNick<br/>:white_check_mark: flrgh<br/>:x: Murillo Paula<br/>:x: HansK-p<br/><hr/>**Murillo Paula** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Kong/lua-resty-healthcheck?pullRequest=144) it.</sub>

nowNick committed:
chore(docs): fix semgrep https warnings

locao commented on rockspecs/lua-resty-healthcheck-0.4.1-2.rockspec:
Why are these rockspecs being removed?

nowNick replied:
They were introduced with:
https://github.com/Kong/lua-resty-healthcheck/commit/7c7e31fa313eefed6f923767ce2a666fb17ced70
and they are not present in 3.0.0. So reverting 2.0.0 version removed those files as well. 
I guess we might actually want to keep them since they point to tar archives that are still reachable... so theoretically someone should be able to download them, right?

locao replied:
These files were added in `master` because they were added in a recent re-release, to avoid issues with conflicting dependencies.

> I guess we might actually want to keep them since they point to tar archives that are still reachable... so theoretically someone should be able to download them, right?

Yes, I think we should keep them.

locao commented on README.md:
```diff
@@ -117,32 +102,13 @@ Versioning is strictly based on [Semantic Versioning](https://semver.org/)
+* Perf: optimize by localizing some functions [#92](https://github.com/Kong/lua-resty-healthcheck/pull/92) (backport)
+* Fix: Generate fresh default http_statuses within new() [#83](https://github.com/Kong/lua-resty-healthcheck/pull/83) (backport)
+
+### 2.0.0
```
It seems to me we can keep the module's history.

nowNick replied:
I've kept the history but added a note to make sure nobody expects changes from 2.0.0 to automatically be in 3.0.0

locao commented on README.md:
```diff
@@ -1,8 +1,7 @@
-![latest version](https://img.shields.io/github/v/tag/Kong/lua-resty-healthcheck?sort=semver)
-![latest luarocks version](https://img.shields.io/luarocks/v/kong/lua-resty-healthcheck?style=flat-square)
-![master branch](https://github.com/Kong/lua-resty-healthcheck/actions/workflows/latest_os.yml/badge.svg)
+![legacy version](https://img.shields.io/luarocks/v/kong/lua-resty-healthcheck/1.6.1-1?style=flat-square)
+![Release 1.6.x](https://github.com/Kong/lua-resty-healthcheck/actions/workflows/build_and_test_with_resty_events.yml/badge.svg?branch=release/1.6.x)
```
```suggestion
![latest version](https://img.shields.io/github/v/tag/Kong/lua-resty-healthcheck?sort=semver)
![latest luarocks version](https://img.shields.io/luarocks/v/kong/lua-resty-healthcheck?style=flat-square)
![master branch](https://github.com/Kong/lua-resty-healthcheck/actions/workflows/latest_os.yml/badge.svg)
```

locao commented on rockspecs/lua-resty-healthcheck-2.0.0-1.rockspec:
I think we can keep this file, we are not wiping out 2.0.0 from the books.

locao commented on t/21-run_locked.t:
Shouldn't this file reside in `t/with_worker-events`?

locao reviewed (changes requested):
Just a few details.

nowNick commented on lua-resty-healthcheck-2.0.0-1.src.rock:
Hmm... it is removed

![Screenshot 2023-11-07 at 15 43 31](https://github.com/Kong/lua-resty-healthcheck/assets/5149215/a4714095-30ac-4be5-8fab-851cf3a53a16)

locao replied:
LOL sorry, my brain must have short-circuited while checking this 🤦

nowNick committed:
docs(readme): update shield badges

Co-authored-by: Vinicius Mignot <vinicius.mignot@gmail.com>

nowNick committed:
chore(*): add 2.0.0 rockspecs and fix tests

Release 2.0.x introduced some rockspecs with fixes.
Reverting back to 1.3.0 and reapplying changes from 3.0.0
reversed those fixes. This commit reintroduces them.

KAG-2704

locao commented on .github/workflows/sast.yml:
```diff
@@ -1,5 +1,10 @@
 name: SAST
 
+concurrency:
+  # for PR's cancel the running task, if another commit is pushed
```
This is great! 🙌

nowNick replied:
Thanks 😊 ! But I need to give credit where credit is due: @Tieske 👏 
https://github.com/Kong/lua-resty-healthcheck/pull/142#discussion_r1358181243

locao reviewed (approved):
Great work! 🎉
